### PR TITLE
Add class `Drawing` and test

### PR DESCRIPTION
**Pull request by realtomkwok**
- Set up Jest
- Write unit tests for each method of the Drawing class

### CODE REVIEW

File: backend/src/__test__/drawing.test.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+				
+				// Check Buffer
+				const buffer = emptyDrawing.getValue()
+				expect(Buffer.isBuffer(buffer)).toBe(true)
+				
+				// Check the size of the drawing
+				expect(buffer.length).toEqual(totalArraySize)
+				
+				// Check if it's empty
+				expect(buffer.every(byte => byte === 0)).toBe(true)
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:24]: I think that this could be a separate test. This test is testing three things, when it could just test one. This means that if any of these assertions fail, this test will fail. I think it would be clearer if, in this case, we just had one assertion for each test case.

Additionally, the first assertion failure would end the test, hiding future assertions from running.

What do you think?

File: backend/src/__test__/drawing.test.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		})
+		
+		// for `canCreate()`
+		test('should validate if a buffer can be used to create a drawing', () => {
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:28]: So here you are testing multiple test cases, but I think it could make sense as a single test case as you are testing a single behaviour—whether it returns a true or false result.

However, I think this test would be clearer if it was written using something like [test.each](https://jestjs.io/docs/api#testeachtablename-fn-timeout). I think this has the added benefit of dynamically creating a separate test case per input.

File: backend/src/drawing.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+export class Drawing {
```
Comment (LivingLimes) [inline backend/src/drawing.ts:1]: Good job moving this to its own file

File: backend/src/index.ts [hunk 4]
```diff
@@ -1,90 +1,44 @@
-import { createServer } from 'http'
```
Comment (LivingLimes) [inline backend/src/index.ts:1]: So it looks like your formatter works differently to mine, which is fine for the new files you've created, but in this file, it has created a lot of diffs that make it hard to review and see what has changed. Can you undo these changes so that only the meaningful changes are diffed?

We can fix the formatting issue later.

File: backend/tsconfig.json [hunk 5]
```diff
@@ -15,6 +15,7 @@
     "allowUnreachableCode": false,
     "noUnusedLocals": true,
     "noUnusedParameters": true,
+    "types": ["node", "jest"],
```
Comment (LivingLimes) [inline backend/tsconfig.json:18]: Can you explain what this line does?

File: backend/.gitignore [hunk 6]
```diff
@@ -27,7 +27,7 @@ lib-cov
 coverage
 *.lcov
 
-# nyc test coverage
+# nyc __test__ coverage
```
Comment (LivingLimes) [inline backend/.gitignore:30]: Hmm I'm wondering why you edited this comment

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,4 @@
 .terraform*
 *.tfstate*
-tfplan
\ No newline at end of file
+tfplan
+/.idea/
```
Comment (LivingLimes) [inline .gitignore:4]: Not that important, but what is this folder?

File: backend/jest.config.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
+/**
+ * For a detailed explanation regarding each configuration property, visit:
+ * https://jestjs.io/docs/configuration
+ */
+
+import type { Config } from "jest"
+
+const config: Config = {
```
Comment (LivingLimes) [inline backend/jest.config.ts:8]: nit: There is a lot going on in this config, but only a few options are in use. Could you reduce it to only the used options as it's difficult to parse right now. 

Since we both haven't really used Jest before, I think it's a good idea to describe in detail what each option does and why the value is set as such if it's not obvious. Writing it out like this would really help you understand and break down what each option does. I'm not going to block this PR because of this, but it was just an idea.

 I've done something like this for the [typescript configuration in this repository](https://github.com/LivingLimes/draw-freely/commit/2fd1cf76ceb4f72ee81888602ea55d152273cac4).

File: backend/jest.config.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+		// setupFilesAfterEnv: [],
+		
+		// The number of seconds after which a __test__ is considered as slow and reported as such in the results.
+		// slowTestThreshold: 5,
+		
+		// A list of paths to snapshot serializer modules Jest should use for snapshot testing
+		// snapshotSerializers: [],
+		
+		// The __test__ environment that will be used for testing
+		testEnvironment: "jest-environment-node",
```
Comment (LivingLimes) [inline backend/jest.config.ts:143]: do you know how this differs from `node`?

File: backend/jest.config.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+		// automock: false,
+		
+		// Stop running tests after `n` failures
+		// bail: 0,
+		
+		// The directory where Jest should store its cached dependency information
+		// cacheDirectory: "/private/var/folders/kt/y0r0pcx938l9pnmjlxpbm8zm0000gn/T/jest_dx",
+		
+		// Automatically clear mock calls, instances, contexts and results before every __test__
+		clearMocks: true,
```
Comment (LivingLimes) [inline backend/jest.config.ts:19]: Hmm why did you set this to true? I feel like this behaviour is a bit unintuitive and since we aren't really mocking yet, I wouldn't set it to true unless we had a good reason to.

File: backend/jest.config.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		// The directory where Jest should output its coverage files
+		coverageDirectory: "coverage",
+		
+		// An array of regexp pattern strings used to skip coverage collection
+		coveragePathIgnorePatterns: [
+		  "/node_modules/"
+		],
+		
+		// Indicates which provider should be used to instrument code for coverage
+		coverageProvider: "v8",
```
Comment (LivingLimes) [inline backend/jest.config.ts:36]: What is this?

File: backend/src/__test__/drawing.test.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+		// Test for the drawing class
+		// for `createEmpty()`
+		test('should crate an empty canvas of the correct size', () => {
+				// Check Drawing instance
+				const emptyDrawing = Drawing.createEmpty()
+				expect(emptyDrawing).toBeInstanceOf(Drawing)
+				
+				// Check Buffer
+				const buffer = emptyDrawing.getValue()
+				expect(Buffer.isBuffer(buffer)).toBe(true)
```
[the remark this answers is not in the document]
Comment (realtomkwok) [reply backend/src/__test__/drawing.test.ts:18]: I was thinking about ensuring what it gets from `getValue()` is a Buffer. But after a second look it seems redundant cuz the function should always return a Buffer

File: backend/src/__test__/drawing.test.ts (hunk 1 above)
Comment (realtomkwok) [reply backend/src/__test__/drawing.test.ts:24]: Yeah makes sense.  I'll be separating them into 3 another tests

File: backend/.gitignore (hunk 6 above)
Comment (realtomkwok) [reply backend/.gitignore:30]: I renamed the directory from `test` to `__test__`

File: backend/tsconfig.json (hunk 5 above)
Comment (realtomkwok) [reply backend/tsconfig.json:18]: Basically it provides the type definition for modules "node" and "jest". I got a bunch of type definition errors before I tried adding this and install the `ts-node` dependency. I can remove this line and test if it still works

File: backend/jest.config.ts (hunk 10 above)
Comment (realtomkwok) [reply backend/jest.config.ts:19]: Yeah that's a mistake while it comes with the initial setup configs

File: backend/jest.config.ts (hunk 11 above)
Comment (realtomkwok) [reply backend/jest.config.ts:36]: It's one of the steps in the setup where you can choose between `v8` and `babel-jest` (by default). Claude says it refers to the tool or library that Jest will use behind the scenes to track which lines of code are executed during tests. And `v8` will be running faster and more accurate than `babel-jest` for Node.js apps

File: backend/jest.config.ts (hunk 9 above)
Comment (realtomkwok) [reply backend/jest.config.ts:143]: They're the same haha

File: backend/.gitignore (hunk 6 above)
Comment (realtomkwok) [reply backend/.gitignore:30]: Ha I knew what was going on! I renamed the directory using "Refactor"  in my IDE but it replaced "test" into "__test__"...

File: backend/tsconfig.json (hunk 5 above)
Comment (LivingLimes) [reply backend/tsconfig.json:18]: I'm looking at the documentation here and it says it restricts access to `@types` packages inside node modules here: https://www.typescriptlang.org/tsconfig/#types

I also see that you've installed `ts-node` as a dependency (https://www.npmjs.com/package/ts-node), but the `tsx` package (https://github.com/privatenumber/tsx) actually does the same thing.

I tried running `npm start` and `npm test` and had no issues after removing the ts-node package and this line. Can you try this again and let me know if that's consistent with what you see?

File: backend/src/__test__/drawing.test.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+const pixelSize = 4
+const totalArraySize = canvasHeight * canvasWidth * pixelSize
+
+describe("test Drawing class", () => {
+		describe("test `createEmpty()`)", () => {
+				const emptyDrawing = Drawing.createEmpty()
+				const buffer = emptyDrawing.getValue()
+				
+				test("should create an instance of Drawing", () => {
+						expect(emptyDrawing).toBeInstanceOf(Drawing)
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:14]: Can you explain what behaviour you're testing and how this test helps build confidence in the code working better?

File: backend/src/__test__/drawing.test.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+						expect(buffer.length).toEqual(totalArraySize)
+				})
+				
+				test("value of the buffer should be empty", () => {
+						expect(buffer.every(el => el >= 0)).toBe(true)
+				})
+		})
+		
+		describe("test `canCreate()", () => {
+				test.each([
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:27]: I think this looks good. I do notice though that 

```js
if (!drawing.every(el => el >= 0 && el <= 255 && Number.isFinite(el))) {
						return false
				}
```

could add to the test cases here!

File: backend/src/__test__/drawing.test.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+						{ testName: "buffer is too small, should return false", buffer: Buffer.alloc(totalArraySize - 1, 0), expected: false },
+						{ testName: "buffer is too large, should return false", buffer: Buffer.alloc(totalArraySize + 1, 0), expected: false },
+						{ testName: "invalid input, should return false", buffer: Object.create(null), expected: false },
+				])("$testName", ({ buffer, expected }) => {
+						expect(Drawing.canCreate(buffer)).toBe(expected)
+				})
+		})
+		
+		describe("test `createFrom()`", () => {
+				const validDrawingBuffer = Drawing.createEmpty().getValue()
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:38]: When it comes to unit testing, I would advise against use one method's output to test another method's input. This is because if `Drawing.createEmpty()` returns a bad output one day, this test will also fail.

I would also consider testing the creation of a drawing that isn't empty just to make sure that works

File: backend/src/__test__/drawing.test.ts [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+				const invalidDrawingBuffer = Buffer.alloc(totalArraySize - 1, 0)
+				
+				test("should create a drawing canvas with the correct values", () => {
+						expect(Drawing.createFrom(validDrawingBuffer)).toBeInstanceOf(Drawing)
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:42]: I really like that your test is named 'should create a drawing canvas with correct values', but your actual assertion only checks the type.

It would be more precise to be checking that the actual drawing value is correct. This is not too different to what you did here in another test:

```js
expect(buffer.every(el => el >= 0)).toBe(true)
```

File: backend/src/__test__/drawing.test.ts [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+				test("should create an instance of Drawing", () => {
+						expect(emptyDrawing).toBeInstanceOf(Drawing)
+				})
+				
+				test("should be the same size as `totalArraySize`", () => {
+						expect(buffer.length).toEqual(totalArraySize)
+				})
+				
+				test("value of the buffer should be empty", () => {
+						expect(buffer.every(el => el >= 0)).toBe(true)
```
Comment (LivingLimes) [inline backend/src/__test__/drawing.test.ts:22]: Nice work here, should this be `el === 0`?

File: backend/src/__test__/README.md [hunk 18]
```diff
@@ -0,0 +1,38 @@
+# Jest
```
Comment (LivingLimes) [inline backend/src/__test__/README.md:1]: Could you move this to the PR description since it's documentation about tooling and can be accessed when finding this commit / PR again.

**Review by LivingLimes — CHANGES_REQUESTED** [review]

**Comment by realtomkwok** [comment]
# About Jest

* [Documentation](https://jestjs.io/docs/configuration#coveragedirectory-string)
* [Guide to Understanding the Jest Coverage Report](https://aakanksha-02.medium.com/understanding-the-jest-coverage-report-a-complete-guide-57d950130fdc#:~:text=Types%20of%20Coverage%20in%20Jest&text=Statements%20Coverage%3A%20This%20shows%20the,statements%20coverage%20would%20be%2080%25.)
* [What's Test Driven Development](https://www.youtube.com/watch?v=Jv2uxzhPFl4): TDD is to write tests first before
  implementing actual code. It goes from **red** (write failing tests), to **green** (write some code to get it passed
  the tests), and to **refactor**, which is to refine the code based on the test results. TDD works great if the project
  has clear requirements and goals.

## Jest configs (`jest.config.ts`)

### Coverage

Coverage report indicates how well the code is tested. If enabled, Jest will generate a coverage report under
`./src/coverage` (set by `coverageDirectory`)

Check [this](https://aakanksha-02.medium.com/understanding-the-jest-coverage-report-a-complete-guide-57d950130fdc#:~:text=Types%20of%20Coverage%20in%20Jest&text=Statements%20Coverage%3A%20This%20shows%20the,statements%20coverage%20would%20be%2080%25.)
on understanding the jest coverage report.

* `collectCoverage`: Whether the test collects coverage information. Enable it for more clarity on the effectiveness of
  the written tests.
* `coverageProvider`: Indicates which provider should be used to instrument code for coverage. `babel`'
  s [istanbul](https://istanbul.js.org) is the default coverage provider. It generally more predictable and stable
  coverage results than `v8`, however, it can be slower and more memory-intensive, especially for large files.
* `coverageThreshold`: This will be used to configure minimum threshold enforcement for coverage results. If the
  threshold aren't met, the test will fail. It can be specified as `global`, and as a directory or a file path.
    * We should always aim for 100% coverage while it may be hard to reach irl

### Test Environment

* `testEnvironment`: The test environment that will be used for testing. The default environment is `node`. Another option is `jsdom` which creates a browser-like environment, and it's suitable for testing web apps.
* `testEnvironmentOption`

### Miscellaneous
* `moduleFileExtensions`: An array of file extensions your modules use. It is recommended to place the extensions most commonly used in the project on the left.
* `transform`: A transformer is needed if the code contains syntax not supported by Node out of the box (such as JSX, TypeScript, etc.). Default transformer is `babel-jest` and it would transform any `.js(x)` and `ts(x)` file.
    * Use `ts-jest` for TypeScript-written projects
* `verbose`: Enable it to have each test reported during the run.
* `watchman`: Enable it to watch files and record when they change. Jest to re-run only the tests that might be affected by those changes.

File: backend/src/__test__/drawing.test.ts (hunk 13 above)
Comment (LivingLimes) [reply backend/src/__test__/drawing.test.ts:14]: So typescript gives us type safety so we can expect this value to be a Drawing. I reckon you can remove the line if you agree

Comment (realtomkwok) [reply backend/src/__test__/drawing.test.ts:14]: Yep fair enough

**Comment by realtomkwok** [comment]
- [ ] Add class `Player` and `Players`: 

```mermaid
classDiagram
    class Player {
        %% Socket id
        string id
        string screenName
        Status status
        +setScreenName(id)
        +getScreenName(id)
        +setStatus
        +checkStatus(status)
    }

    class Players {
        List players
        Player currentPlayer
        +addPlayer()
        +removePlayer()
        +getPlayer()
        +getNextPlayer(id) Player
        +setTurnPlayer()
        +getTurnPlayer()
        +rotateTurn()
    }

    class Status {
        <<enumeration>>
        Online
        %% User is in the room but game hasn't started (e.g. looking for players, etc)
        Pending
        Playing
        Offline
    }

    class GameMode {
        <<enumeration>>
        OneLine
        LineLengthLimit
    }
```

File: backend/src/index.ts [hunk 19]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
 let turnPlayer: string | undefined | null = undefined
```
Comment (realtomkwok) [inline backend/src/index.ts:48]: Replace `players` with an instance of `Players`

File: backend/src/index.ts [hunk 20]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
 
 io.on(SOCKET_EVENTS_INBOUND.CONNECTION, (socket) => {
-  console.log('User connected:', socket.id);
-
-  players.push(socket.id)
```
Comment (realtomkwok) [inline backend/src/index.ts:99]: Replace it with using class method `addPlayer()`

File: backend/src/index.ts [hunk 21]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
 let turnPlayer: string | undefined | null = undefined
 
 io.on(SOCKET_EVENTS_INBOUND.CONNECTION, (socket) => {
-  console.log('User connected:', socket.id);
-
```
Comment (realtomkwok) [inline backend/src/index.ts:98]: Add a new player by using `Players.addPlayer()`

File: backend/src/index.ts [hunk 22]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
-  players.push(socket.id)
-
-  if (players.length === 1) {
-    turnPlayer = socket.id
-  }
-
-  socket.emit(SOCKET_EVENTS_OUTBOUND.UPDATE_TURN, {
-    turnPlayer
```
Comment (realtomkwok) [inline backend/src/index.ts:106]: Replace this with using class method `Players.updateTurn()`

File: backend/src/index.ts [hunk 23]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
-    socket.broadcast.emit(SOCKET_EVENTS_OUTBOUND.SELECTED_GAME_MODE, mode)
-  })
-
-  socket.on(SOCKET_EVENTS_INBOUND.CLEAR_CANVAS, () => {
-    if (!players.includes(socket.id)) {
-      return
-    }
-
-    drawing = Drawing.createEmpty()
-    gameMode = null
```
Comment (realtomkwok) [inline backend/src/index.ts:136]: Just wondering why would clearing the canvas change its game mode as well?

File: backend/src/index.ts [hunk 24]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
-      gameMode
-    })
-  })
-
-  socket.on(SOCKET_EVENTS_INBOUND.END_TURN, () => {
-    if (turnPlayer !== socket.id || !players.includes(socket.id)) {
-      return
-    }
-
-    const nextPlayer = getNextElement(players, turnPlayer)
```
Comment (realtomkwok) [inline backend/src/index.ts:148]: Replace this with class method `Players.getNextPlayer()`

File: backend/src/index.ts [hunk 25]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
-    console.log('User disconnected', socket.id, reason)
-
-    // Turn player disconnect special case
-    if (turnPlayer === socket.id) {
-      const nextPlayer = getNextElement(players, turnPlayer)
-      turnPlayer = nextPlayer ?? players[0] ?? null
-    }
-
-    // Turn player and viewer disconnect
-    players = players.filter(p => p !== socket.id)
```
Comment (realtomkwok) [inline backend/src/index.ts:166]: Replace this with method `Players.removePlayer()`

File: backend/src/index.ts [hunk 26]
```diff
@@ -94,84 +48,84 @@ const lineLengthLimit: number = 150
...
-    })
-  })
-
-  socket.on(SOCKET_EVENTS_INBOUND.END_TURN, () => {
-    if (turnPlayer !== socket.id || !players.includes(socket.id)) {
-      return
-    }
-
-    const nextPlayer = getNextElement(players, turnPlayer)
-    turnPlayer = nextPlayer ?? players[0] ?? null
```
Comment (realtomkwok) [inline backend/src/index.ts:149]: Replace this with using `Players.setTurnPlayer()`

**Review by realtomkwok — COMMENTED** [review]
A [more accurate](https://stackoverflow.com/a/29972322) timer than using `setInterval` or `setTimeout`. It can be used to:
* Set a timer for each round
* Return an accurate remaining time

## Usage

```ts
import Timer from "./timer"

// new Timer(duration: number, tickInterval?: number)

// Create a 10-second timer, whose interval between each tick lasts 1 second
const timer = new Timer(10000, 1000)

// Start the timer with given callback function after each tick and when time's up
timer.start(onTickFn, onCompleteFn)

// Stop the timer
timer.stop()

// Reset the timer
timer.reset()

// Check if the timer is running
timer.isRunning()

// Get the remaining time
timer.remainingTime
```

## Methods
* `start(onTick, onComplete)`: Start the timer and pass the optional `onTick` and `onComplete` function if given.
  * `onTick(remainingTime)`: _optional_. If given, the callback function will be called after each tick, which can be set by parameter `tickInterval` when initialising the timer.
  * `onComplete()`: _optional_. The given function will be called when the timer finishes
* `stop() => void`: If there's a timer running (`intervalId === true`), use this method to stop it.
* `reset()`: Stop the timer and reset its remaining time to the given duration
* `isRunning()`: Check if the timer is running
* `timer.remainingTime`: return remaining milliseconds of the running timer

## Tests
The following test cases have checked and passed:
* The timer should start properly
* The timer should stop properly
* The timer should be reset properly
* The timer should return the correct remaining time
* Callback functions should be called in after each tick and/or times up
